src/modules/audio/audio.c

```diff
@@ -195,68 +180,91 @@ bool lovrAudioInit(AudioConfig config[2]) {
+      if (strcmp(playbackDevices[i].name, state.config[AUDIO_PLAYBACK].deviceName) == 0) {
+        config.playback.pDeviceID = &playbackDevices[i].id;
+      }
     }
-
-    // .. and create new one
-    if (state.config[i].enable) {
-      lovrAudioInitDevice(i);
+    if (state.config[AUDIO_PLAYBACK].deviceName && config.playback.pDeviceID == NULL) {
+      lovrLog(LOG_WARN, "audio", "No audio playback device called '%s'; falling back to default.", state.config[AUDIO_PLAYBACK].deviceName);
+    }
+    config.playback.channels = OUTPUT_CHANNELS;
+  } else { // if AUDIO_CAPTURE
+    ma_device_type deviceType = ma_device_type_capture;
+    config = ma_device_config_init(deviceType);
+
+    config.capture.format = miniAudioFormatFromLovr[state.config[AUDIO_CAPTURE].format];
+    for(int i = 0; i < captureDeviceCount && state.config[AUDIO_CAPTURE].deviceName; i++) {
+      if (strcmp(captureDevices[i].name, state.config[AUDIO_CAPTURE].deviceName) == 0) {
+        config.capture.pDeviceID = &playbackDevices[i].id;
+      }
+    }
+    if (state.config[AUDIO_CAPTURE].deviceName && config.capture.pDeviceID == NULL) {
+      lovrLog(LOG_WARN, "audio", "No audio capture device called '%s'; falling back to default.", state.config[AUDIO_CAPTURE].deviceName);
     }
+    config.capture.channels = CAPTURE_CHANNELS;
   }
+  config.performanceProfile = ma_performance_profile_low_latency;
+  config.dataCallback = callbacks[type];
+  config.sampleRate = state.config[type].sampleRate;
 
+
+  ma_result err = ma_device_init(&state.context, &config, &state.devices[type]);
+  if (err != MA_SUCCESS) {
+    lovrLog(LOG_WARN, "audio", "Failed to enable %s audio device: %s (%d)\n", type == AUDIO_PLAYBACK ? "playback" : "capture", ma_result_description(err), err);
+    return false;
+  }
   return true;
 }
 
 bool lovrAudioStart(AudioType type) {
-  if (state.config[type].enable == false) {
-    if (lovrAudioInitDevice(type) == false) {
-      if (type == AUDIO_CAPTURE) {
-        lovrPlatformRequestPermission(AUDIO_CAPTURE_PERMISSION);
-        // lovrAudioStart will be retried from boot.lua upon permission granted event
-      }
-      return false;
-    }
-    state.config[type].enable = state.config[type].start = true;
+  bool initResult = lovrAudioInitDevice(type);
+  if (initResult == false && type == AUDIO_CAPTURE) {
+    lovrPlatformRequestPermission(AUDIO_CAPTURE_PERMISSION);
+    // lovrAudioStart will be retried from boot.lua upon permission granted event
+    return false;
   }
-  int status = ma_device_start(&state.devices[type]);
+  ma_result status = ma_device_start(&state.devices[type]);
   return status == MA_SUCCESS;
 }
 
 bool lovrAudioStop(AudioType type) {
-  return ma_device_stop(&state.devices[type]) == MA_SUCCESS;
+  ma_result stoppingResult = ma_device_stop(&state.devices[type]);
+  ma_device_uninit(&state.devices[type]);
+
+  return stoppingResult == MA_SUCCESS;
 }
```

Comment:
Is there a reason why start/stop init/destroy the devices?  It would be nice if `lovr.audio.start` and `lovr.audio.stop` could be used as an interface to ma_device_start and ma_device_stop (I am assuming this is more lightweight than completely tearing down and recreating the devices).  The create/destroy could still happen as part of useDevice.

Reply:
@bjornbytes Yes there's a reason, see the comment in [this resolved conversation](https://github.com/bjornbytes/lovr/pull/347#discussion_r555748132). In short: reducing the amount of state and complexity. Having both "loaded" and "started" was feeling messy. But maybe I overdid it. I could just remove destroy from stop, and add a destroy to useDevice and module deinit, and I think that would be equivalent to what you're suggesting. That would make it impossible to deinit an audio device without quitting lovr or useDevice'ing, but maybe that's fine. What do you think?

Reply:
Ah sorry I missed that.  I would be okay with any of these 3:

- Removing start and stop
- Having start/stop initialize/destroy and start/stop the device
- Having start/stop start/stop the device

After thinking about it I still like #3 the most because the ability to pause a device without destroying it seems useful.

Removing reset is fine, reset only existed to let people switch to the default audio device after it changed, but with an explicit useDevice function for it reset is no longer needed.

I think the conf.lua flags for enable/start could still be useful, but I don't mind leaving them out until it's clear they're needed.

Reply:
Okay, I have implemented 2 now. I will attempt to remove init/deinit from start/stop so it becomes 3.

Reply:
@bjornbytes Done! Hope this PR is now to your satisfaction :)